Fix deserialization of Instance(HasProps) and add HasProps.__eq__

mattpap opened the pull request:
fixes #5118

bryevdv commented:
Appears to be a py2/3 issue with these changes. Is #5118 the correct issue BTW? I don't immediately see the connection.

mattpap commented on bokeh/core/properties.py:
```diff
@@ -1231,7 +1243,7 @@ def from_json(self, json, models=None):
 
                 for name, value in iteritems(json):
-                    prop = self.instance_type.lookup(name)
+                    prop = self.instance_type.lookup(name).descriptor
```
@bryevdv, yes. This line appears in the traceback in the issue. The remainder of changes is to allow `assertEquals()` to work.

mattpap commented:
> TypeError: unhashable type: 'BoxAnnotation'

That's correct. Currently `HasProps` (and `Model`) implement `__eq__` and `__ne__`, but don't implement `__hash__`, because that would conflict with mutability of models. We assume hashability here and there, e.g. by constructing sets of models. One way to deal with this to make `Model` hashable by id and use only id in equality testing. Object will be still mutable, but that won't affect `==` and hash and their relationship. Though `HasProps` would have to stay unhashable.

bryevdv commented:
@mattpap that seems OK

mattpap commented:
> that seems OK

No, because it defeats the whole purpose of allowing to structurally compare models, which I need for testing and interactive sessions (well, that's pretty much testing again). To get structural equality working across the board, I removed `__eq__` and implemented `equals()` that just compares properties and their values, and disregards `id`.

I think we can implement `__eq__` in general, but it won't be structural equality, especially that we want also mutability and hashability. What has to be done is to make `HasProps` have `id` field (move from `Model`). This is how bokehjs works and I think that's the correct approach. This will unify both versions. Note that we already implicitly assume this, by allowing Python to use intrinsic object identification (memory location) to be used for hashability.

Also, I would like to rename `HasProps` to `Struct`, because it's hard to refer meaningfully to `HasProps` in writing.

mattpap commented:
How naive of me:
```
File "/home/travis/build/bokeh/bokeh/bokeh/core/tests/test_properties.py", line 1505
    v = 1.equals(Foo(x=1))
         ^
SyntaxError: invalid syntax
```
assuming that Python is a regular programming language.

bryevdv commented:
`int(1).equals(Foo(x=1)` might work I'm not sure if it tests something you care about. 

I'm +0 on renaming to `Struct`

I think `HasProps` having an intrinsic `id` is a bit weird, since I associate `id` specifically with things that can be and are serialized, i.e. `Model` and subclasses. I'm not necessarily opposed tho, but I would have to really dig in to things a bit to really evaluate. I'd say make the PR you want then let's review it in detail.

mattpap commented:
@bryevdv, I will continue discussion on `__eq__` and `__hash__` elsewhere.
